src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -845,48 +853,143 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
+  QString name;
+  QgsLayerTreeNode *node = index2node( index );
+  QgsLayerTreeModelLegendNode *ltmln = index2legendNode( index ); // Possibly useless
+  QgsLayerTreeLayer *nodeLayer = QgsLayerTree::isLayer( node ) ? QgsLayerTree::toLayer( node ) : nullptr;
+  if ( nodeLayer )
+  {
+    QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
+    if ( !ltmln && ( role == Qt::DisplayRole || role == Qt::EditRole ) )
     {
-      QgsLayerTreeLayer *nodeLayer = QgsLayerTree::toLayer( node );
-      QString name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
-      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() && role == Qt::DisplayRole )
+      //finding the first label that is stored
+      name = nodeLayer->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = nodeLayer->name();
+      if ( name.isEmpty() )
+        name = node->customProperty( QStringLiteral( "legend/title-label" ) ).toString();
+      if ( name.isEmpty() )
+        name = node->name();
+
+      if ( nodeLayer->customProperty( QStringLiteral( "showFeatureCount" ), 0 ).toInt() )
       {
-        QgsVectorLayer *vlayer = qobject_cast<QgsVectorLayer *>( nodeLayer->layer() );
         if ( vlayer && vlayer->featureCount() >= 0 )
+        {
           name += QStringLiteral( " [%1]" ).arg( vlayer->featureCount() );
+          Q_UNUSED( ltmln );
+          return name;
+        }
+      }
+    }
+    if ( ( nodeLayer || ltmln ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) )
+    {
+      QgsExpressionContext context = ( mLayoutLegendContext ) ? QgsExpressionContext( *mLayoutLegendContext ) : QgsExpressionContext();
+      if ( vlayer )
+      {
+        connect( vlayer, &QgsVectorLayer::startCount, this, &QgsLegendModel::pendingCount );
+        connect( vlayer, &QgsVectorLayer::countDone, this, &QgsLegendModel::doneCount );
+      }
+
+      if ( ltmln )
+      {
+        if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( ltmln ) )
+          name = synode->evaluateLabel( context ); // removed name input; existing symbol/model tree have distinct names
+        return name;
+      }
+      else
+      {
+        const QList<QgsLayerTreeModelLegendNode *> legendnodes = layerLegendNodes( nodeLayer, false );
+        if ( legendnodes.count() > 1 ) // evaluate all existing legend nodes but leave the name for the legend evaluator
+        {
+          for ( QgsLayerTreeModelLegendNode *treenode : legendnodes )
+          {
+            if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( treenode ) )
+              synode->evaluateLabel( context );
+          }
+        }
+        else if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( legendnodes.first() ) )
+          name = synode->evaluateLabel( context, name );
```

Comment:
@m-kuhn I noticed you are patching dynamic_cast, would it be a good idea for me to make a new commit to convert those to qobject_cast?

Reply:
Yes, whenever possible prefer `qobject_cast` over `dynamic_cast`